packages/runtime/src/store/tutorial-runner.ts

```diff
@@ -503,6 +552,77 @@ export class TutorialRunner {
     this._updateDirtyState(files);
   }
 
+  private _stopWatcher(): void {
+    // if there was a watcher terminate it
+    if (this._watcher) {
+      this._watcher.close();
+      this._watcher = undefined;
+    }
+  }
+
+  private _setupWatcher(webcontainer: WebContainer) {
+    // inform that the watcher could be installed if we wanted to
+    this._readyToWatch = true;
+
+    // if the watcher is alreay setup or we don't sync content we exit
+    if (this._watcher || !this._watchContentFromWebContainer) {
+      return;
+    }
+
+    const filesToRead = new Map<string, 'utf-8' | null>();
+
+    let timeoutId: ReturnType<typeof setTimeout> | undefined;
+
+    const readFiles = () => {
+      const files = [...filesToRead.entries()];
+
+      filesToRead.clear();
+
+      Promise.all(
+        files.map(async ([filePath, encoding]) => {
+          // casts could be removed with an `if` but it feels weird
+          const content = (await webcontainer.fs.readFile(filePath, encoding as any)) as Uint8Array | string;
+
+          return [filePath, content] as const;
+        }),
+      ).then((fileContents) => {
+        for (const [filePath, content] of fileContents) {
+          this._editorStore.updateFile(filePath, content);
+        }
+      });
+    };
+
+    const scheduleReadFor = (filePath: string, encoding: 'utf-8' | null) => {
+      filesToRead.set(filePath, encoding);
+
+      clearTimeout(timeoutId);
+      timeoutId = setTimeout(readFiles, 100);
+    };
```

Comment:
Could we add comment here stating what this does, something like `"// add file to read-queue and discard previously queued read task`. 

I had to check manually that this doesn't accidentally cancel the previous reads if multiple ones are done within 100ms. 

Really clever and compact scheduler btw 💯

Reply:
Definitely! Very good point! Thanks for checking it :smiley: